### PR TITLE
Feature/game logic

### PR DESCRIPTION
**Pull request by JAAR91**
Add game logic to the project

Now is possible to play, the games consist of 2 players, each has a true not mark a number on a board, that one that 
makes a row of 3 symbols win.

The project includes the game logic, an input checker to avoid crashes due to invalid inputs or empty inputs, also the menu is complete and this includes the instructions and the game credits.

### CODE REVIEW

File: bin/credits.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    puts '+----------------------------------------------------------------------------------+'
+    puts '|       Thank you for playing the game!                                            |'
+    puts '|            v                                                                     |'
+    puts '|                                                                                  |'
+    puts '|          (°v°)                                           BY:                     |'
+    puts '|         /|-=-|\                                          JAAR                    |'
+    puts '|         ( _=_ )                                          ERI                     |'
+    puts '|          ^^ ^^                                                        MICROVERSE |'
+    puts '+----------------------------------------------------------------------------------+'
+    puts 'Press any key to go back to menu....'
```
Comment (mricanho) [inline bin/credits.rb]: You can't use puts outside main.rb 😢

File: bin/credits.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  def bigtictactou
+    puts '+----------------------------------------------------------------------------------+'
+    puts '|   lllllllllllllll       ]llllllllllllllL           }lllllllllllllj               |'
+    puts '|   lllll||||||llll       Jllll|l||||||llL           lllll||||l|llll               |'
+    puts '|        ||||l  www  ,wyww,    l||||L  wwwwww   ,wwww,    L||||  ,wwww,  ;wwwww,   |'
+    puts '|        l|||l  lll ||||jl|j   ll|||L  ||l||l  l|lrl||L   ||||l jlll*|ll ||||***   |'
+    puts '|        |||||  l|l ||||       ll|||L ;||ll||L ||lL       L|||l ]||l ||l ||||ll    |'
+    puts '|        |||||  l|l ||||J$$%   l||||L $|ll$||l l|lL$$$L   L|||l ]||l,||l ||||,,,   |'
+    puts '|        YWWWW  WWW  "4&l&*`   YWWWWL WWW""WW&  *&&lW*    YWWWW  lY&l&M  !WWWWWF   |'
+    puts '+----------------------------------------------------------------------------------+'
```
Comment (mricanho) [inline bin/credits.rb]: You can't use puts outside main.rb 😢

File: bin/displayprint.rb [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    puts "|#{input}|"
+    print '+'
+    m_line.each { |item| print item }
+    puts '+'
+    puts '   v'
+    puts ''
+    puts array[0]
+    puts array[1]
+    puts array[2]
+    puts array[3]
```
Comment (mricanho) [inline bin/displayprint.rb]: You can't use puts or prints outside main.rb 😢

File: bin/displayprint.rb [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    array[1].each { |item| print item }
+    puts ''
+    puts array[2]
+    array[3].each { |item| print item }
+    puts ''
+    puts array[4]
+    array[5].each { |item| print item }
+    puts ''
+    puts array[6]
+    puts ''
```
Comment (mricanho) [inline bin/displayprint.rb]: You can't use puts outside main.rb 😢

File: bin/game.rb [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  def take_turns
+    game = 0
+    @current_player = @first_player if @current_player.nil?
+
+    while game.zero?
+      @display.display_clear
+      @display.board_print(@board)
+      puts '+-----------------------------------------------------------+'
+      puts "It is #{@current_player} turn:"
+      puts 'Please select any valid spot or cell on the board to play'
```
Comment (mricanho) [inline bin/game.rb]: You can't use puts outside main.rb 😢

File: bin/game.rb [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    end
+    end_game_message(game)
+    end_game
+  end
+
+  def end_game
+    @display.board_print(@board)
+    puts 'Dou you want to start a new game?(Y/N):'
+    puts '1. Yes'
+    puts '2. No'
```
Comment (mricanho) [inline bin/game.rb]: You can't use puts outside main.rb 😢

File: bin/game.rb [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    return game if @board[2] == @board[4] && @board[4] == @board[6]
+
+    return 3 if @board.none?(Integer)
+
+    0
+  end
+
+  def player_turns_done(asnw, symbol)
+    while @board.none?(asnw)
+      puts 'The location selected is not available, please select another one:'
```
Comment (mricanho) [inline bin/game.rb]: You can't use puts outside main.rb 😢

File: bin/gamerules.rb [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+class Gamerules
+  def initialize
+    @display = Display.new
+  end
+
+  def show_game_rules
+    @display.display_clear
+    @display.jaarix('Game Rules!!!!')
+    puts 'Press enter to continue....'
```
Comment (mricanho) [inline bin/gamerules.rb]: You can't use puts outside main.rb 😢

File: bin/gamerules.rb [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  def rules
+    @display.display_clear
+    @display.jaarix('The game consist on 2 players switching turn and marking a place on a 3x3 board!')
+    puts 'Press enter to continue.....'
+    gets
+  end
+
+  def win_one
+    @display.display_clear
+    puts 'Page 1 of 5'
```
Comment (mricanho) [inline bin/gamerules.rb]: You can't use puts outside main.rb 😢

File: bin/gamerules.rb [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    @display.board_print(['X', 'X', 'X', 4, 5, 6, 7, 8, 9])
+    @display.board_print([1, 2, 3, 'O', 'O', 'O', 7, 8, 9])
+    @display.board_print([1, 2, 3, 4, 5, 6, 'O', 'O', 'O'])
+    puts 'Press enter to continue.....'
+    gets
+  end
+
+  def win_three
+    @display.display_clear
+    puts 'Page 3 of 5'
```
Comment (mricanho) [inline bin/gamerules.rb]: You can't use puts outside main.rb 😢

File: bin/inputchecker.rb [hunk 11]
```diff
@@ -0,0 +1,17 @@
+class Inputchecker
+  def number_checker(menu_input, initial, last)
+    until !menu_input.nil? && menu_input.is_a?(Integer) && menu_input >= initial && menu_input <= last
+      print "Please enter only numbers beetwen #{initial} and #{last} please!! :"
```
Comment (mricanho) [inline bin/inputchecker.rb]: You can't use prints outside main.rb 😢

File: bin/inputchecker.rb [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+      menu_input = gets.chomp.to_i
+    end
+    menu_input
+  end
+
+  def player_name_check(input)
+    while input.nil? || input.length.zero?
+      puts 'Please enter a valid name, empty value not accepted'
```
Comment (mricanho) [inline bin/inputchecker.rb]: You can't use puts outside main.rb 😢

File: bin/menu.rb [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+  end
+
+  def menu_text
+    puts '+------------------------------------+'
+    puts '|------->Menu(Enter a number)<-------|'
+    puts '|1.- Start Game                      |'
+    puts '|2.- Instructions                    |'
+    puts '|3.- Credits                         |'
+    puts '|4.- Exit                            |'
+    puts '+------------------------------------+'
```
Comment (mricanho) [inline bin/menu.rb]: You can't use puts outside main.rb 😢

File: bin/menu.rb [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+  def call_game
+    display = Display.new
+    display.display_clear
+    display.jaarix('Lets Get Started!!')
+    puts '+----------------------------------------------+'
+    puts 'Player Number one please enter your name :'
+    inputchecker = Inputchecker.new
+    player_one = inputchecker.player_name_check(gets.chomp)
+    puts '+----------------------------------------------+'
+    puts 'Player Number two please enter your name :'
```
Comment (mricanho) [inline bin/menu.rb]: You can't use puts outside main.rb 😢

File: bin/game.rb [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+      input_checker = Inputchecker.new
+      asnw = input_checker.number_checker(gets.chomp.to_i, 1, 9)
```
Comment (mricanho) [inline bin/game.rb]: You can't use gets.chomp outside main.rb 😢

**Review by mricanho — CHANGES_REQUESTED** [review]
# 🔴  Changes Required 🔴  

Hello team✋🏽!  You did a great job so far with your Tic-Tac-Toe🔥 the behavior of the game is very good and the interface super clear, you are using the single responsibility rule for your methods in an incredible way⭐, the README file is good too, but now you need to do a few changes:

- **All game-logic related resources (classes and/or modules) must be kept in the lib directory**  [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png)
- **You have puts, prints, and gets.chomp's outside the main.rb file, this is a no-no, please, fix this point. Remember that main.rb is the only one responsible for user interaction** [screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png)

You can contact me on slack if you have any questions just search for me as @ Miguel Ricaño.

### 💻 👩🏽‍💻  Happy Coding! 🖥️ 👨🏽‍💻

**Comment by rukundoeric** [comment]
# Status: ♻️ CHANGES REQUIRED ♻️
_Great job guys so far 👏 💯 👍. However, there are few issues we need to address before we move on, please check the comments below._

- ### To Highlight
  - [x] The project is clean. 💯
  - [x] The project follows Git flow. 💯
  - [x] The project follows Git flow. 💯
  - [x] UI looks nice 💯

- ### To Improve  
  - [x] [MANDATORY] `main.rb` is not executable. from the instruction of this project, I should be able to start the game by simply running `bin/main.rb` without specifying `ruby` keyword.
  - [x] [MANDATORY] `displayprint.rb and game.rb` files should be moved from `bin/` dir to `lib/` dir.

_Please make all the required changes before you submit another code review. If you have any problems or concerns, you can reach out to me on Slack: Rukundo Eric 🚀 . I wish you all the very best 🥇_

### _Keep working hard._
### _Happy Coding😃_
-----------------------------------------------------------------------------------------------------------------------------------------
**Regards,
[Rukundo Eric](https://rukundo-eric.herokuapp.com/) ,
[LinkedIn](https://www.linkedin.com/in/rukundo-eric/), 
Slack: Rukundo Eric**

**Review by henatan99 — APPROVED** [review]
# Project Approved :100: 
## Congratulations @JAAR91 and @errea :fireworks: :partying_face: 
You did an amazing job on this project. 

## To Highlight 
- Your project is superbly user interactive and user-friendly:clap: :clap: :clap: 
- You made all the requested changes by the previous reviewers 
- The game behaves exactly as intended 

You can make use of the below optional comments (while not required for project submission)

[OPTIONAL] 
- You can take the many methods in the main file to dedicated module and class files and call them from the main file using assigned variables for puts string literals as arguments. 

### Enjoy your Coding :smiley: 

You can reach out to me via : 
- [LinkedIn](https://www.linkedin.com/in/henok-mekonnen-2a251613/)
- [Twitter](https://twitter.com/henatan99)
- or Henok Mekonnen on Slack

**Pull request merged**